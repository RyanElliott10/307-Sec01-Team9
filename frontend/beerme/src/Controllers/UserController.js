import NetClient from "../Utils/NetClient";

class UserController {
  static firstName = "";
  static lastName = "";
  static email = "";
  static password = "";
  static isBusiness = true; // change to default of false once live
  static businessName = "";
  static currBeer = "";
  static currABV = "";
  static currIBU = "";
  static currStyle = "";
  static currBeerId = 0;
  static userId = 0;

  static isLoggedIn = false;
  static cachedBeers = [];

  static async login(email, password) {
    console.log(`UserController: login ${email} ${password}`);
    this.email = email;
    this.password = password;

    const data = {
      email: email,
      password: password
    };

    return NetClient.post("https://localhost:44300/api/login", data).then(
      data => {
        if (data) {
          this.userId = data.Id;
          this.firstName = data.Name.split(" ")[0];
          this.lastName = data.Name.split(" ")[1];
          this.email = data.Email;
          this.isBusiness = data.IsBusiness;
          this.businessName = data.BusinessName;
          this.isLoggedIn = true;
          return this.isLoggedIn;
        } else {
          this.isLoggedIn = false;
          return this.isLoggedIn;
        }
      }
    );
  }

  static async createAccount(
    firstName,
    lastName,
    email,
    password,
    isBusiness = false,
    businessName = ""
  ) {
    this.email = email;
    this.password = password;

    const data = {
      name: `${firstName} ${lastName}`,
      email: email,
      password: password,
      isBusiness: isBusiness,
      businessName: businessName
    };

    return NetClient.post("https://localhost:44300/api/users", data).then(
      data => {
        if (data) {
          this.userId = data.Id;
          this.firstName = data.Name.split(" ")[0];
          this.lastName = data.Name.split(" ")[1];
          this.email = data.Email;
          this.isBusiness = data.IsBusiness;
          this.businessName = data.BusinessName;
          this.isLoggedIn = true;
          return this.isLoggedIn;
        } else {
          this.isLoggedIn = false;
          return this.isLoggedIn;
        }
      }
    );
  }

  static getCurrentUserObject() {
    return {
      Id: this.userId,
      Name: `${this.firstName} ${this.lastName}`,
      Email: this.email,
      Password: this.password,
      IsBusiness: this.isBusiness,
      BusinessName: this.businessName
    };
  }

  static getCurrentUser() {
    return this.isLoggedIn;
  }

  static async fetchAllBeers() {
    NetClient.get("https://localhost:44300/api/beers").then(data => {
      this.cachedBeers = data;
    });
  }
}

export default UserController;
