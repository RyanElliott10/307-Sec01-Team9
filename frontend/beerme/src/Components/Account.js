import React, { Component } from "react";
import { Button, Form, Col } from "react-bootstrap";
import Select from "react-select";
import { Redirect } from "react-router-dom";

import NetClient from "../Utils/NetClient";
import UserController from "../Controllers/UserController";

// FOR POST:
//  BeerName, StyleId

export class Account extends Component {
  constructor(props) {
    super(props);
    this.state = {
      addBeerName: "",
      addBeerData: {
        label: "",
        value: null
      }
    };
  }

  componentDidMount() {
    this.setState({
      allStyles: [{ value: 0, label: "test" }]
    });
    NetClient.get("https://localhost:44300/api/BeerStyles").then(data => {
      const styles = data.map(d => {
        return {
          label: d.Style,
          value: d.Id
        };
      });

      this.setState({
        allStyles: styles
      });
    });
  }

  _validateAddBeerForm() {
    return (
      this.state.addBeerData.label.length > 0 &&
      this.state.addBeerData.value !== null &&
      this.state.addBeerName.length > 0
    );
  }

  _validateRemoveBeer() {
    return true;
  }

  _handleAddBeerSubmit = async event => {
    event.preventDefault();
    NetClient.post("https://localhost:44300/api/Beers", {
      BeerName: this.state.addBeerName,
      StyleId: this.state.addBeerData.value
    });
    this.setState(
      {
        addBeerName: "",
        addBeerData: {
          label: "",
          value: null
        }
      },
      () => (this.addBeerFormRef.value = this.state.addBeerName)
    );
  };

  _handleRemoveBeerSubmit = async event => {
    event.preventDefault();
  };

  _renderControl(type, value, isDisabled, onChange = () => {}) {
    return (
      <Form.Control
        disabled={isDisabled}
        type={type}
        placeholder={value}
        onChange={onChange}
      />
    );
  }

  _renderBusinessName() {
    if (UserController.isBusiness) {
      return <h2>{UserController.businessName}</h2>;
    }
    return null;
  }

  _renderCommon() {
    return (
      <Form.Group as={Col} controlId="formGridFirstName">
        <Form onSubmit={this._handleCreateAccountSubmit}>
          <Form.Row>
            <Form.Group as={Col} controlId="formGridFirstName">
              <Form.Label>First Name</Form.Label>
              {this._renderControl("name", UserController.firstName, true)}
            </Form.Group>
            <Form.Group as={Col} controlId="formGridLastName">
              <Form.Label>Last Name</Form.Label>
              {this._renderControl("name", UserController.lastName, true)}
            </Form.Group>
          </Form.Row>

          <Form.Row>
            <Form.Group as={Col} controlId="formGridEmail">
              <Form.Label>Email Address</Form.Label>
              {this._renderControl("email", UserController.email, true)}
            </Form.Group>
          </Form.Row>
        </Form>
      </Form.Group>
    );
  }

  _renderButton(variant, disabled, onClick, text) {
    return (
      <Button
        type="submit"
        variant={variant}
        disabled={disabled}
        onClick={onClick}
      >
        {text}
      </Button>
    );
  }

  _renderAddBeerNameField() {
    return (
      <>
        <Form.Label>Name of Beer</Form.Label>
        <Form.Control
          ref={form => (this.addBeerFormRef = form)}
          disabled={false}
          type={"name"}
          placeholder={"Corona"}
          onChange={e => {
            const newName = e.target.value;
            this.setState({
              addBeerName: newName
            });
          }}
        />
      </>
    );
  }

  _renderAddBeerDropdown() {
    return (
      <>
        <Form>
          <Form.Label>Style</Form.Label>
        </Form>
        <Select
          options={this.state.allStyles}
          value={this.state.addBeerData}
          onChange={event => {
            this.setState({
              addBeerData: {
                label: event.label,
                value: event.value
              }
            });
          }}
        />{" "}
      </>
    );
  }

  _renderAddBeer() {
    return (
      <Form.Group as={Col} controlId="formGridFirstName">
        <h2>Add a Beer</h2>
        <hr />
        <Form.Row>
          <Form.Group as={Col} controlId="formGridBeerName">
            {this._renderAddBeerNameField()}
          </Form.Group>
        </Form.Row>
        {this._renderAddBeerDropdown()}
        <hr />
        <Form.Row>
          {this._renderButton(
            "primary",
            !this._validateAddBeerForm(),
            this._handleAddBeerSubmit,
            "Add Beer"
          )}
        </Form.Row>
      </Form.Group>
    );
  }

  _renderRemoveBeer() {
    return (
      <Form.Group as={Col} controlId="formGridLastName">
        <h2>Remove a Beer</h2>
        <hr />
        <Form.Label>Name of Beer</Form.Label>
        <Select
          options={this.state.allStyles}
          onChange={event =>
            this.setState({
              addBeerData: {
                name: this.state.addBeerData.name,
                value: event.value
              }
            })
          }
        />
        <hr />
        {this._renderButton(
          "danger",
          !this._validateRemoveBeer(),
          this._handleRemoveBeerSubmit,
          "Remove Beer"
        )}
      </Form.Group>
    );
  }

  _renderAddRemove() {
    if (!UserController.isBusiness) {
      return null;
    }

    return (
      <Form.Group as={Col} controlId="formGridFirstName">
        <Form onSubmit={this._handleCreateAccountSubmit}>
          <Form.Row>
            {this._renderAddBeer()}
            <div />
            {this._renderRemoveBeer()}
          </Form.Row>
        </Form>
      </Form.Group>
    );
  }

  render() {
    if (UserController.getCurrentUser()) {
      return (
        <div className="Login" style={broadStyle}>
          <Form>
            {this._renderBusinessName()}
            {this._renderCommon()}
            {this._renderAddRemove()}
          </Form>
        </div>
      );
    }
    return <Redirect to="/account-entry" />;
  }
}

const broadStyle = {
  paddingTop: "50px",
  marginLeft: "100px",
  marginRight: "100px",
  alignItems: "center",
  justifyContent: "center"
};

export default Account;
