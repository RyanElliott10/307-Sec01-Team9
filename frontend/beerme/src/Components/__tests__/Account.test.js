import React from "react";
import renderer from "react-test-renderer";
import { shallow } from "enzyme";
import { BrowserRouter } from "react-router-dom";

import Account from "../Account";
import UserController from "../../Controllers/UserController";

it("renders correctly", () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <Account />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with a user logged in", () => {
  UserController.isLoggedIn = true;
  const tree = renderer
    .create(
      <BrowserRouter>
        <Account />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with a business", () => {
  UserController.isLoggedIn = true;
  UserController.isBusiness = true;
  UserController.businessName = "Test Business";
  const tree = renderer
    .create(
      <BrowserRouter>
        <Account />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with state", () => {
  const ref = shallow(
    <BrowserRouter>
      <Account />
    </BrowserRouter>
  ).instance();
  ref.setState({
    addBeerData: {
      name: "Corona",
      styles: "Style-here",
      color: "Pale"
    }
  });

  const expectedState = {
    addBeerData: {
      name: "Corona",
      styles: "Style-here",
      color: "Pale"
    }
  };

  expect(ref.state).toEqual(expectedState);
});
