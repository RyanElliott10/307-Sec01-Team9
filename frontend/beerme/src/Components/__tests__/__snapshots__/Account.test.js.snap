// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `null`;

exports[`renders correctly with a business 1`] = `
<div
  className="Login"
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "marginLeft": "100px",
      "marginRight": "100px",
      "paddingTop": "50px",
    }
  }
>
  <form
    className=""
  >
    <h2>
      Test Business
    </h2>
    <div
      className="form-group col"
    >
      <form
        className=""
      >
        <div
          className="form-row"
        >
          <div
            className="form-group col"
          >
            <label
              className="form-label"
              htmlFor="formGridFirstName"
            >
              First Name
            </label>
            <input
              className="form-control"
              disabled={true}
              id="formGridFirstName"
              onChange={[Function]}
              placeholder=""
              type="name"
            />
          </div>
          <div
            className="form-group col"
          >
            <label
              className="form-label"
              htmlFor="formGridLastName"
            >
              Last Name
            </label>
            <input
              className="form-control"
              disabled={true}
              id="formGridLastName"
              onChange={[Function]}
              placeholder=""
              type="name"
            />
          </div>
        </div>
        <div
          className="form-row"
        >
          <div
            className="form-group col"
          >
            <label
              className="form-label"
              htmlFor="formGridEmail"
            >
              Email Address
            </label>
            <input
              className="form-control"
              disabled={true}
              id="formGridEmail"
              onChange={[Function]}
              placeholder=""
              type="email"
            />
          </div>
        </div>
      </form>
    </div>
    <div
      className="form-group col"
    >
      <form
        className=""
      >
        <div
          className="form-row"
        >
          <div
            className="form-group col"
          >
            <h2>
              Add a Beer
            </h2>
            <hr />
            <div
              className="form-row"
            >
              <div
                className="form-group col"
              >
                <label
                  className="form-label"
                  htmlFor="formGridBeerName"
                >
                  Name of Beer
                </label>
                <input
                  className="form-control"
                  disabled={false}
                  id="formGridBeerName"
                  onChange={[Function]}
                  placeholder="Corona"
                  type="name"
                />
              </div>
            </div>
            <form
              className=""
            >
              <label
                className="form-label"
                htmlFor="formGridFirstName"
              >
                Style
              </label>
            </form>
            <div
              className=" css-2b097c-container"
              onKeyDown={[Function]}
            >
              <div
                className=" css-yk16xz-control"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <div
                  className=" css-1hwfws3"
                >
                  <div
                    className=" css-1uccc91-singleValue"
                  >
                    
                  </div>
                  <div
                    className="css-b8ldur-Input"
                  >
                    <div
                      className=""
                      style={
                        Object {
                          "display": "inline-block",
                        }
                      }
                    >
                      <input
                        aria-autocomplete="list"
                        autoCapitalize="none"
                        autoComplete="off"
                        autoCorrect="off"
                        disabled={false}
                        id="react-select-2-input"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        spellCheck="false"
                        style={
                          Object {
                            "background": 0,
                            "border": 0,
                            "boxSizing": "content-box",
                            "color": "inherit",
                            "fontSize": "inherit",
                            "label": "input",
                            "opacity": 1,
                            "outline": 0,
                            "padding": 0,
                            "width": "1px",
                          }
                        }
                        tabIndex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style={
                          Object {
                            "height": 0,
                            "left": 0,
                            "overflow": "scroll",
                            "position": "absolute",
                            "top": 0,
                            "visibility": "hidden",
                            "whiteSpace": "pre",
                          }
                        }
                      >
                        
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  className=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    className=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    className=" css-tlfecz-indicatorContainer"
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <svg
                      aria-hidden="true"
                      className="css-6q0nyr-Svg"
                      focusable="false"
                      height={20}
                      viewBox="0 0 20 20"
                      width={20}
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <hr />
            <div
              className="form-row"
            >
              <button
                className="btn btn-primary"
                disabled={true}
                onClick={[Function]}
                type="submit"
              >
                Add Beer
              </button>
            </div>
          </div>
          <div />
          <div
            className="form-group col"
          >
            <h2>
              Remove a Beer
            </h2>
            <hr />
            <label
              className="form-label"
              htmlFor="formGridLastName"
            >
              Name of Beer
            </label>
            <div
              className=" css-2b097c-container"
              onKeyDown={[Function]}
            >
              <div
                className=" css-yk16xz-control"
                onMouseDown={[Function]}
                onTouchEnd={[Function]}
              >
                <div
                  className=" css-1hwfws3"
                >
                  <div
                    className=" css-1wa3eu0-placeholder"
                  >
                    Select...
                  </div>
                  <div
                    className="css-b8ldur-Input"
                  >
                    <div
                      className=""
                      style={
                        Object {
                          "display": "inline-block",
                        }
                      }
                    >
                      <input
                        aria-autocomplete="list"
                        autoCapitalize="none"
                        autoComplete="off"
                        autoCorrect="off"
                        disabled={false}
                        id="react-select-3-input"
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        spellCheck="false"
                        style={
                          Object {
                            "background": 0,
                            "border": 0,
                            "boxSizing": "content-box",
                            "color": "inherit",
                            "fontSize": "inherit",
                            "label": "input",
                            "opacity": 1,
                            "outline": 0,
                            "padding": 0,
                            "width": "1px",
                          }
                        }
                        tabIndex="0"
                        type="text"
                        value=""
                      />
                      <div
                        style={
                          Object {
                            "height": 0,
                            "left": 0,
                            "overflow": "scroll",
                            "position": "absolute",
                            "top": 0,
                            "visibility": "hidden",
                            "whiteSpace": "pre",
                          }
                        }
                      >
                        
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  className=" css-1hb7zxy-IndicatorsContainer"
                >
                  <span
                    className=" css-1okebmr-indicatorSeparator"
                  />
                  <div
                    aria-hidden="true"
                    className=" css-tlfecz-indicatorContainer"
                    onMouseDown={[Function]}
                    onTouchEnd={[Function]}
                  >
                    <svg
                      aria-hidden="true"
                      className="css-6q0nyr-Svg"
                      focusable="false"
                      height={20}
                      viewBox="0 0 20 20"
                      width={20}
                    >
                      <path
                        d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
            </div>
            <hr />
            <button
              className="btn btn-danger"
              disabled={false}
              onClick={[Function]}
              type="submit"
            >
              Remove Beer
            </button>
          </div>
        </div>
      </form>
    </div>
    <form
      className=""
    >
      <button
        className="btn btn-danger"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Logout
      </button>
    </form>
  </form>
</div>
`;

exports[`renders correctly with a user logged in 1`] = `
<div
  className="Login"
  style={
    Object {
      "alignItems": "center",
      "justifyContent": "center",
      "marginLeft": "100px",
      "marginRight": "100px",
      "paddingTop": "50px",
    }
  }
>
  <form
    className=""
  >
    <div
      className="form-group col"
    >
      <form
        className=""
      >
        <div
          className="form-row"
        >
          <div
            className="form-group col"
          >
            <label
              className="form-label"
              htmlFor="formGridFirstName"
            >
              First Name
            </label>
            <input
              className="form-control"
              disabled={true}
              id="formGridFirstName"
              onChange={[Function]}
              placeholder=""
              type="name"
            />
          </div>
          <div
            className="form-group col"
          >
            <label
              className="form-label"
              htmlFor="formGridLastName"
            >
              Last Name
            </label>
            <input
              className="form-control"
              disabled={true}
              id="formGridLastName"
              onChange={[Function]}
              placeholder=""
              type="name"
            />
          </div>
        </div>
        <div
          className="form-row"
        >
          <div
            className="form-group col"
          >
            <label
              className="form-label"
              htmlFor="formGridEmail"
            >
              Email Address
            </label>
            <input
              className="form-control"
              disabled={true}
              id="formGridEmail"
              onChange={[Function]}
              placeholder=""
              type="email"
            />
          </div>
        </div>
      </form>
    </div>
    <form
      className=""
    >
      <button
        className="btn btn-danger"
        disabled={false}
        onClick={[Function]}
        type="submit"
      >
        Logout
      </button>
    </form>
  </form>
</div>
`;
